valhalla/thor/unidirectional_astar.h

```diff
@@ -185,8 +185,9 @@ class UnidirectionalAStar : public PathAlgorithm {
   // Edge status. Mark edges that are in adjacency list or settled.
   EdgeStatus edgestatus_;
 
-  // Destinations, id and percent used along the edge
-  std::unordered_map<uint64_t, float> destinations_percent_along_;
+  // Mark if edge is a destination
+  std::unordered_multimap<valhalla::baldr::GraphId, std::reference_wrapper<const valhalla::PathEdge>>
```

Comment:
im not too sur that we need a multimap here. the correlated edges can only have one copy of each unique edge id so you wont ever push in the same edge twice as a destination. unless im missing something here.

Reply:
That's not correct. The map-matcher could match the same edge multiple times. Just think about a hard loop shaped like a U with the car between.

Reply:
while it is true that map matching needs to handle this case, *this* algorithm has nothing to do with our map matching implementation. all our map matching code is in the `meili` namespace (and a little bit of it in thor/map_matcher thor/trace_*). this is because we do a one-to-many routing between map matching candidates and this algorithm only does a one-to-one routing.

Reply:
I mean if the incoming destination set has the same edge multiple times in the list, then we would need to add complex logic for tie breaking. Using a multimap usually nicely handles that case without any other checks required.

Reply:
yep sorry what im saying is that that cannot happen. the list is populated by this line: https://github.com/valhalla/valhalla/blob/4532f7be1910349bb30ccdad36fb2e9c1d930881/src/thor/unidirectional_astar.cc#L822

that list in turn is populated by loki search and im just telling you that loki search will only have unique edge ids in the list. this is because the whole job of loki search is to find the closest point on a given edge and return that as a candidate. there can be only 1 closest point on a given edge per location. so im just saying that multimap is an unneeded complication. i can see you are using it to nicely handle the event that there would be the same id more than once but what im saying is that there will never be.

Reply:
One should not couple implementation details of two algorithms, esp. across public APIs.

So relying on the fact that LOKI behaves that way is not good because the A* could also be used in a different way.

Reply:
one should probably also not constantly change the interface to those public APIs but i guess it depends on the guarantees one wants to provide with a given API.

since its inception the project has not really concerned itself with maintaining compatibility (commit to commit) between the APIs exposed in the headers. we pretty much only gaurantee the json API. in fact it is a long term goal to vastly reduce the scope of methods exposed in the library headers (a battle for another day.. or maybe slowly plugging away at it rather).

furthermore, i can tell you for a fact that other parts of thor assume that there are no duplicates in the list of a locations edge candidates (eg: https://github.com/valhalla/valhalla/blob/master/src/thor/triplegbuilder.cc#L411 if there were duplicates here it would randomly pick one candidate, not necessarily the right one)

anyway the whole reason i was telling you this was because i saw the `multimap` and i saw the call to `equal_range`  and i was wondering: why is he protecting against this case that doesn't occur? which lead to me really just trying to do a polite: FYI you dont need to worry about this, it cant happen. but i think from your round about answers afterward i could understand that you may be calling this algorithm with your own set of edge candidates which may indeed have duplicates?

my point is i understand how one *could* put more than one correlation point for the same edge in the list, it just doesnt happen in any of the code here on github. if you have other designs elsewhere or future plans, best to just say that and shortcut the discussion :smile:

Reply:
Nay, no plans right now, but could happen.

I did that change for the simple reason of my edge-cases-tuned brain :-)